### ISSUE TITLE
Check for the merge inflection point inside the git history before using it for comparison

### ISSUE THREAD

**Issue by orta** [danger/danger#480]
https://circleci.com/gh/danger/danger.systems/89

I hadn't merged master in to this branch, so assumably the midway point - noted in https://github.com/danger/danger/pull/432 isn't available in the git history for the branch. We should detect this and do  `git fetch origin master` to grab some extra commits on the CI side.

**Comment by cysp** [comment]
I was able to reproduce this, with a PR that was raised from one-commit-behind-`HEAD` of the target branch.
The commit that git complained about not being able to find was the current `HEAD` of the target branch.

It seems to me that the problem is that we're using the sha that Github (in this case) tells us is the head of the target (`base`) branch but the CI system (in my case, Buildkite) has not fetched that target branch so it does not have that commit locally.

**Comment by cysp** [comment]
I'm inclined to fix this in the CI, by either getting it to fetch the `merge` commit for the PR or both the PR's head and the target branch's head.
(Though I understand that this might not be a possible solution for all CI services. 😝 )

**Comment by kylejm** [comment]
I found that this error is still occurring. With the help for @orta, I added these lines:

```
self.scm.exec "fetch" if self.scm.exec("--no-pager show #{base_commit}").empty?
self.scm.exec "fetch" if self.scm.exec("--no-pager show #{head_commit}").empty?
```

[here](https://github.com/danger/danger/blob/master/lib/danger/request_source/github.rb#L56), which fixed the issue.

I'm still wrapping my head around what's going wrong (never having written Ruby before isn't helping). I'm guessing that this call to fetch from the repo isn't explicitly being made early enough, before this call?
